docs/rfcs/0009-remove-interactors-and-controllers.md

```diff
@@ -188,6 +188,150 @@ This should simplify the search for underlying logic by:
+
+To demonstrate the bullets above, here is the method definition in the `DefaultPocketStoriesController` that currently handles the business logic initiated from the `interactor::onCategoryClicked` call above. 
+
+```
+    override fun handleCategoryClick(categoryClicked: PocketRecommendedStoriesCategory) {
+        val initialCategoriesSelections = appStore.state.pocketStoriesCategoriesSelections
+
+        // First check whether the category is clicked to be deselected.
+        if (initialCategoriesSelections.map { it.name }.contains(categoryClicked.name)) {
+            appStore.dispatch(AppAction.DeselectPocketStoriesCategory(categoryClicked.name))
+            Pocket.homeRecsCategoryClicked.record(
+                Pocket.HomeRecsCategoryClickedExtra(
+                    categoryName = categoryClicked.name,
+                    newState = "deselected",
+                    selectedTotal = initialCategoriesSelections.size.toString(),
+                ),
+            )
+            return
+        }
+
+        // If a new category is clicked to be selected:
+        // Ensure the number of categories selected at a time is capped.
+        val oldestCategoryToDeselect =
+            if (initialCategoriesSelections.size == POCKET_CATEGORIES_SELECTED_AT_A_TIME_COUNT) {
+                initialCategoriesSelections.minByOrNull { it.selectionTimestamp }
+            } else {
+                null
+            }
+        oldestCategoryToDeselect?.let {
+            appStore.dispatch(AppAction.DeselectPocketStoriesCategory(it.name))
+        }
+
+        // Finally update the selection.
+        appStore.dispatch(AppAction.SelectPocketStoriesCategory(categoryClicked.name))
+
+        Pocket.homeRecsCategoryClicked.record(
+            Pocket.HomeRecsCategoryClickedExtra(
+                categoryName = categoryClicked.name,
+                newState = "selected",
+                selectedTotal = initialCategoriesSelections.size.toString(),
+            ),
+        )
+    }
+```
+
+If we break this down into a pseudocode algorithm, we get the following steps:
+1. Read the current state from the store
+2. If the category is being deselected
+    1. Dispatch an action to update the state
+    2. Record a metric that the category was deselected
+    3. Return early
+3. If the number of categories selected would be over max
+    1. Dispatch an action deselecting the oldest
+4. Dispatch an action with the newly selected category
+5. Record a metric that a category was selected
+
+In order to transform this algorithm into an appropriate usage of the Redux pattern, we can separate this list into two parts: steps that have side-effects and steps which can be represented by pure state updates. In this case, only the steps that record metrics represent side-effects. This will mean that those steps should be moved into a middleware, and the rest can be moved into a reducer.
+
+Here's how that might look:
+```kotlin
```

Comment:
From [my original comment](https://github.com/mozilla-mobile/firefox-android/pull/1466#discussion_r1170798740):
> Here, you have made a change to replace the interactor with the dispatch call. Can you add an example of what the other end of the store looks like that connects to this change?
> 
> I'm trying to visualize how would features like bookmarks connect to the storage and the snackbar UI layer, or navigate.

I might have not been clear, sorry. It wasn't the redux/store part that is hard to visualize. It's the other end of the store where we now have a state change and we need to tell the UI (for example) to update. How will that work?

In the example of Pocket you've used, it falls under the Compose extensions that observe state changes. What about Android View code? We have plenty of that, so how will we observe and react to state changes there? We should be adding an example of that connecting piece more than the internals of how to use `lib-state` which we have plenty of examples in our code base.

If you're looking for ideas: I've tried tackling this problem by using individual [`AbstractBinding`](https://github.com/mozilla-mobile/firefox-android/blob/main/android-components/components/lib/state/src/main/java/mozilla/components/lib/state/helpers/AbstractBinding.kt) classes which observe, collect, and update which ever system needs to know about it. [`TabCounterBinding`](https://github.com/mozilla-mobile/firefox-android/blob/46f3836ae308bd5f050a226c1857363810e62000/fenix/app/src/main/java/org/mozilla/fenix/tabstray/TabCounterBinding.kt) is one of many other "binding" classes in the tabs tray package. What I like about these types of classes are that they are small and easy to grok, testable, and not particularly tied to be part of the UI; use them for anything.

Reply:
Ah gotcha, thanks for clarifying. On the one hand, it almost seems like if we are investing in the code health of a feature by refactoring it for architectural consistency it would also make sense to me to update that area to Compose - but that probably won't match reality in every case. I'll work up a View example as well

Reply:
> ..to update that area to Compose - but that probably won't match reality in every case.

Not all state is tied to UI, it's just one example of something that needs to react to a state change. We can have other systems that need to observe state changes and do other work. Having all of those in middlewares do not always make sense too.

I would use the tabs tray as [another](https://github.com/mozilla-mobile/firefox-android/blob/main/fenix/app/src/main/java/org/mozilla/fenix/tabstray/browser/BrowserTabsAdapter.kt#L61) [example](https://github.com/mozilla-mobile/firefox-android/blob/main/fenix/app/src/main/java/org/mozilla/fenix/tabstray/browser/AbstractBrowserTrayList.kt#L35-L37), where we need to observe and react to state but it doesn't always mean a middleware or constructor dependency makes sense.

Reply:
> Not all state is tied to UI, it's just one example of something that needs to react to a state change.

While true, I think the context of thinking about this in terms of interactors/controllers tends to frame it as more of a discussion of UI state.

> We can have other systems that need to observe state changes and do other work. Having all of those in middlewares do not always make sense too.

If we need to react to state changes, that means observing them from somewhere. We could have free-floating global components that observe for the lifetime of the app, or we could have middleware react to the changes. Otherwise, even if they aren't necessarily tied to the UI these components seem like they would at least be tied to the lifecycle of UI components.

> I would use the tabs tray as [another](https://github.com/mozilla-mobile/firefox-android/blob/910afa889ebc5222a1b9a877837de5c55244d441/fenix/app/src/main/java/org/mozilla/fenix/tabstray/browser/BrowserTabsAdapter.kt#L61) [example](https://github.com/mozilla-mobile/firefox-android/blob/910afa889ebc5222a1b9a877837de5c55244d441/fenix/app/src/main/java/org/mozilla/fenix/tabstray/browser/AbstractBrowserTrayList.kt#L35-L37), where we need to observe and react to state but it doesn't always mean a middleware or constructor dependency makes sense.

These are good examples, thank you! But I assume this statement was unrelated to the rest of the comment, since this is UI code? Were you thinking these should just be included as examples in the doc?

Reply:
Yeah, fair points. Maybe lets assume they are all UI components: you have Compose covered, so maybe add an example how an observation to update and Android View would work, and that should be enough. 🙂

Reply:
I see you went with using the `AbstractBinding` direction for connecting Android Views with store observations - that's fair. FWIW, just stating it in this doc like the Compose example would have sufficed. 🙂